src/golang/lib/workflow/artifact/artifact.go

```diff
@@ -0,0 +1,184 @@
+type Artifact interface {
+	ID() uuid.UUID
+	Type() artifact.Type
+	Name() string
+
+	// Computed indicates whether this artifact has been computed or not. An artifact is
+	// only considered "computed" if the operator that generates it has completed
+	// successfully.
+	Computed(ctx context.Context) bool
+
+	// PersistResult writes the data of this artifact to a backing store so it can be
+	// fetched later.
+	// Errors if the artifact has not yet been computed.
+	PersistResult(ctx context.Context, opStatus shared.ExecutionStatus) error
+
+	// Finish is an end-of-lifecycle hook meant to do any final cleanup work.
+	Finish(ctx context.Context)
+
+	// GetMetadata fetches the metadata for this artifact.
+	// Errors if the artifact has not yet been computed.
+	GetMetadata(ctx context.Context) (*artifact_result.Metadata, error)
+
+	// GetContent fetches the content of this artifact.
+	// Errors if the artifact has not yet been computed.
+	GetContent(ctx context.Context) ([]byte, error)
+}
+
+func initializeArtifactResultInDatabase(
+	ctx context.Context,
+	artifactID uuid.UUID,
+	workflowDagResultID uuid.UUID,
+	artifactResultWriter artifact_result.Writer,
+	contentPath string,
+	db database.Database,
+) (uuid.UUID, error) {
+	artifactResult, err := artifactResultWriter.CreateArtifactResult(
+		ctx,
+		workflowDagResultID,
+		artifactID,
+		contentPath,
+		db,
+	)
+	if err != nil {
+		return uuid.Nil, errors.Wrap(err, "Failed to create operator result record.")
+	}
+	return artifactResult.Id, nil
+}
+
+type ArtifactImpl struct {
+	id           uuid.UUID
+	name         string
+	description  string
+	artifactType artifact.Type
+
+	contentPath  string
+	metadataPath string
+
+	artifactResultWriter artifact_result.Writer
+	artifactResultID     uuid.UUID
+
+	storageConfig *shared.StorageConfig
+	db            database.Database
+
+	resultsPersisted bool
+}
+
+func (a *ArtifactImpl) ID() uuid.UUID {
+	return a.id
+}
+
+func (a *ArtifactImpl) Type() artifact.Type {
+	return a.artifactType
+}
+
+func (a *ArtifactImpl) Name() string {
+	return a.name
+}
+
+func (a *ArtifactImpl) Computed(ctx context.Context) bool {
+	// An artifact is only considered computed if its metadata path has been populated.
+	return utils.CheckIfObjectExistsInStorage(
+		ctx,
+		a.storageConfig,
+		a.metadataPath,
+	)
+}
+
+func (a *ArtifactImpl) PersistResult(ctx context.Context, opStatus shared.ExecutionStatus) error {
+	if a.resultsPersisted {
+		return errors.Newf("Artifact %s was already persisted!", a.name)
+	}
+	if !a.Computed(ctx) {
+		return errors.Newf("Artifact %s cannot be persisted because it has not been computed.", a.name)
+	}
+	utils.UpdateArtifactResultAfterComputation(
+		ctx,
+		opStatus,
+		a.storageConfig,
+		a.metadataPath,
+		a.artifactResultWriter,
+		a.artifactResultID,
+		a.db,
+	)
+	a.resultsPersisted = true
+	return nil
+}
+
+func (a *ArtifactImpl) Finish(ctx context.Context) {
+	utils.CleanupStorageFile(ctx, a.storageConfig, a.metadataPath)
```

Comment:
Sounds good, I'll make a task for it since this PR is already giant. https://linear.app/aqueducthq/issue/ENG-1363/reorganize-workflowutilsgo-into-more-specific-packages